models/turbine_models/custom_models/sdxl_inference/unet.py

```diff
@@ -70,9 +72,101 @@ def forward(
+        repo_id = hf_model_name
+        subfolder = "unet"
+        revision = "76d28af79639c28a79fa5c6c6468febd3490a37e"
+
+    def download(filename):
+        return hf_hub_download(
+            repo_id=repo_id, subfolder=subfolder, filename=filename, revision=revision
+        )
+
+    results = {
+        "config.json": download("config.json"),
+        "params.safetensors": download("params.safetensors"),
+    }
+    if precision == "i8":
+        results["quant_params.json"] = download("quant_params.json")
+        output_path = external_weight_path.split("unet")[0] + "punet_dataset_i8.irpa"
+        ds = get_punet_i8_dataset(
+            results["config.json"],
+            results["quant_params.json"],
+            results["params.safetensors"],
+            output_path,
+            base_params=None,
+        )
+    else:
+        ds = None  # get_punet_dataset(results["config.json"], results["params.safetensors"], base_params=None)
+
+    cond_unet = punet_unet.from_dataset(ds)
+    mdl = CFGPunetModel(cond_unet)
+    return mdl
+
+
+def get_punet_i8_dataset(
+    config_json_path,
+    quant_params_path,
+    params_path,
+    output_path="./punet_dataset_i8.irpa",
+    quant_params_struct=None,
+    base_params=None,
+):
+    from sharktank.models.punet.tools.import_brevitas_dataset import (
+        _load_json,
+        _load_theta,
+        _get_dataset_props,
+        apply_per_layer_quant,
+        Dataset,
+        Theta,
+        InferenceTensor,
+    )
+
+    # Construct the pre-transform dataset.
+    dataset_props = _get_dataset_props(_load_json(config_json_path))
+    quant_params_struct = _load_json(quant_params_path)
+    with safetensors.safe_open(params_path, framework="pt", device="cpu") as st:
+        quant_theta = _load_theta(st)
+    base_theta = None
+    if base_params is not None:
+        print("Initializing from base parameters:", args.base_params)
+        with safetensors.safe_open(base_params, framework="pt", device="cpu") as st:
+            base_theta = _load_theta(st)
+
+    ds = Dataset(dataset_props, quant_theta if base_theta is None else base_theta)
+
+    # The quant_params_struct has quantization parameter structs keyed by full
+    # layer name. We process each of these in turn to produce a per-layer
+    # quantization scheme where no quantized tensors escape their layer.
+    updated_tensors: dict[str, InferenceTensor] = {}
+    for layer_name, qp in quant_params_struct.items():
+        print(f"Applying per-layer quants: {layer_name}")
+        apply_per_layer_quant(quant_theta, layer_name, qp, updated_tensors)
+
+    # Apply updates into a new Theta.
+    theta = base_theta if base_theta is not None else quant_theta
+    flat_tensors = theta.flatten()
+    flat_tensors.update(updated_tensors)
+    ds.root_theta = Theta(flat_tensors)
+
+    # TODO: Post-process to introduce fused cross-layer connections.
+
+    ds.save(output_path, io_report_callback=print)
```

Comment:
There is a bug in the Dataset handling code where if you construct a Dataset on the fly and then use that without loading it from disk, it does not annotate the imported tensors, resulting in them being inlined vs referenced. I need to fix that bug, but the workflow in this case shouldn't be doing this anyway.

If you follow the advice in the comment above and use the main() entrypoint to generate the IRPA on disk, then do Dataset.load(irpa_path), that will work.